### PR TITLE
[Driver][SYCL] Properly pass object contents with FPGA archives

### PR DESCRIPTION
**Pull request by mdtoguchi**
The recent commit https://github.com/intel/llvm/commit/2df7d3f048c0b65d7aebc365b498370b0806aeab introduced an issue when passing along the host objects in the early (AOCR) archive and consumed by the image (AOCX) archive. Package up (wrap) the images instead of just the list of images by using the -batch functionality with the clang-offload-wrapper.

### CODE REVIEW

File: clang/lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -7446,8 +7446,27 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
       if (Args.hasArg(options::OPT_fsycl_link_EQ) &&
           Args.hasArg(options::OPT_fintelfpga)) {
         // Wrap the object when creating an FPGA AOCX or AOCR binary.
-        auto *BC = C.MakeAction<OffloadWrapperJobAction>(LI, types::TY_LLVM_BC);
-        auto *ASM = C.MakeAction<BackendJobAction>(BC, types::TY_PP_Asm);
+        // When the input file is an AOCR (early) archive, the unbundled host
+        // binary consists of a list of objects. We cannot directly wrap that
+        // binary to be consumed later - this has to go through each listed
+        // object.
+        bool FPGAEarly = true;
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:7453]: Should `FPGAEarly` be `true` by default?

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -7446,8 +7446,27 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
...
+        if (auto *A = C.getInputArgs().getLastArg(options::OPT_fsycl_link_EQ))
+          FPGAEarly = (A->getValue() != StringRef("image"));
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:7455]: Any reason why it is not  `A->getValue() == StringRef("early")`

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -10283,6 +10283,11 @@ void OffloadWrapper::ConstructJob(Compilation &C, const JobAction &JA,
     CmdArgs.push_back(
         TCArgs.MakeArgString(Twine("-target=") + Triple.getTriple()));
 
+    if (Inputs[0].getType() == types::TY_Tempfiletable ||
+        Inputs[0].getType() == types::TY_Tempfilelist)
+      // wrapper actual input files are passed via the batch job file table:
```
Comment (srividya-sundaram) [inline clang/lib/Driver/ToolChains/Clang.cpp:10288]: `wrapper actual input files...`
This sentence seems odd to me. Can you clarify?

File: clang/lib/Driver/Driver.cpp [hunk 4]
```diff
@@ -7446,8 +7446,27 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
       if (Args.hasArg(options::OPT_fsycl_link_EQ) &&
           Args.hasArg(options::OPT_fintelfpga)) {
         // Wrap the object when creating an FPGA AOCX or AOCR binary.
-        auto *BC = C.MakeAction<OffloadWrapperJobAction>(LI, types::TY_LLVM_BC);
-        auto *ASM = C.MakeAction<BackendJobAction>(BC, types::TY_PP_Asm);
+        // When the input file is an AOCR (early) archive, the unbundled host
+        // binary consists of a list of objects. We cannot directly wrap that
+        // binary to be consumed later - this has to go through each listed
+        // object.
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:7452]: `We cannot directly wrap that binary to be consumed later - this has to go through each listed object.`

For my understanding: what does "each listed object" refer to here?

File: clang/lib/Driver/Driver.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:7455]: No particular reason - stream of consciousness programming.  I'll update as the reference to the Boolean makes more sense if checking for "early"

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 3 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:10288]: yeah, this doesn't read very well.  It should just say 'Input files are passed via the batch job file table'

File: clang/lib/Driver/Driver.cpp (hunk 4 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:7452]: The file that is being 'wrapped' here is a list of files represented by a `TY_Tempfilelist` file.    The contents of that file is what we are actually wrapping (which is a list of objects)

**Review by srividya-sundaram — APPROVED** [review]

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, this looks ready to merge, can you take a look?  Thanks!

**Pull request merged**